### ISSUE TITLE
Can't view collection content in ReadOnly PropertyGrid

### ISSUE THREAD

**Issue by xceedsoftware** [xceedsoftware/wpftoolkit#1090]
<b>BoucherS[CodePlex]</b> <br />Based on discussion 
https://wpftoolkit.codeplex.com/discussions/660443

**Comment by xceedsoftware** [comment]
<b>BoucherS[CodePlex]</b> <br />Hi, 

This will be fixed in v3.4. 

――――
Get more controls, features, updates and technical support with 
Xceed Toolkit Plus for WPF

**Comment by fragmer** [comment]
I am using 3.5 and the PropertyGrid still does not show collection content — I tried string[], List<string>, IReadOnlyList<string>, and ReadOnlyCollection<string> — none worked.

**Comment by fragmer** [comment]
This is probably a duplicate of #724 and possibly #599 — though either way, the problem does not appear to be addressed in current versions of the Toolkit.

**Comment by XceedBoucherS** [comment]
Hi,

This actually works in Toolkit v3.4 and up, as stated in the release notes : https://xceed.com/release-notes/#toolkit-wpf-3.4.17280.14430

When the PropertyGrid is ReadOnly and contains any property types mentionned, the Collection button can be pressed to view the property content.

ex :  <xctk:PropertyGrid SelectedObject="{Binding}" 
                       IsReadOnly="True"/>



  public partial class MainWindow : Window
  {
    public MainWindow()
    {
      InitializeComponent();

      this.DataContext = new MyObject()
      {
        FirstName = "Bob",
        Data = new string[] { "A", "B", "C" },
        Names = new List<string>() { "One", "Two", "Three" },
        ROCol = new ReadOnlyCollection<string>( new List<string>() { "Mike", "Kelly", "Gina" } )
      };
    }
  }

  public class MyObject
  {
    public string FirstName
    {
      get;
      set;
    }

    public string[] Data
    {
      get;
      set;
    }

    public List<string> Names
    {
      get;
      set;
    }

    public ReadOnlyCollection<string> ROCol
    {
      get;
      set;
    }
  }

**Comment by mstefarov** [comment]
Retesting this with toolkit 4.0.1, and I see that collection content _does_ show in read-only PropertyGrids, but only as long as the property has a public setter:

![2020-10-08_165400 WpfApp1](https://user-images.githubusercontent.com/587809/95524695-e825ef80-0986-11eb-9bac-671515782ce8.png)

But if the property does not have a public setter, there is no way to view collection content:

![2020-10-08_165314 WpfApp1](https://user-images.githubusercontent.com/587809/95524752-11df1680-0987-11eb-9f30-1f7d99e29b91.png)

I am not sure why a public setter is required to view a _read-only collection_ in a _read-only PropertyGrid_.

```cs
    public class IHaveOnlyGetters {
        public string[] Data { get; } = new string[] { "A", "B", "C" };
        public List<string> Names { get; } = new List<string>() { "One", "Two", "Three" };
        public ReadOnlyCollection<string> ROCol { get; } = new ReadOnlyCollection<string>(new List<string>() { "Mike", "Kelly", "Gina" });
    }

    public class IHaveSettersToo {
        public string[] Data { get; set; } = new string[] { "A", "B", "C" };
        public List<string> Names { get; set; } = new List<string>() { "One", "Two", "Three" };
        public ReadOnlyCollection<string> ROCol { get; set; } = new ReadOnlyCollection<string>(new List<string>() { "Mike", "Kelly", "Gina" });
    }
```

**Comment by XceedBoucherS** [comment]
Hi,
Good catch. Thank you for reporting this. 
It will be fixed in the next release v4.1.

**Comment by mstefarov** [comment]
@XceedBoucherS I retested with 4.6.97 and 5.0.103 and the issue described in https://github.com/xceedsoftware/wpftoolkit/issues/1090#issuecomment-705888339 is still reproducible.

**Comment by ben-jobs** [comment]
unsubscribe.
















At 2023-09-19 20:28:24, "XceedBoucherS" ***@***.***> wrote:

Closed #1090 as completed.
